storm-core/src/jvm/org/apache/storm/daemon/GrouperFactory.java

```diff
@@ -0,0 +1,110 @@
+        final Random random = new Random();
+        final List<Integer> targetTasks = Ordering.natural().sortedCopy(unsortedTargetTasks);
+        final int numTasks = targetTasks.size();
+        switch (Thrift.groupingType(thriftGrouping)) {
+            case FIELDS:
+                if (Thrift.isGlobalGrouping(thriftGrouping)) {
+                    return new Grouper() {
+                        @Override
+                        public List<Integer> outTasks(Integer taskId, List value, LoadMapping load) {
+                            if (targetTasks.isEmpty()) {
+                                return null;
+                            }
+                            // It's possible for target to have multiple tasks if it reads multiple sources
+                            return Collections.singletonList(targetTasks.get(0));
+                        }
+                    };
+                }
+                return new Grouper.FieldsGrouper(outFields, thriftGrouping, targetTasks);
+            case SHUFFLE:
+                return new Grouper.ShuffleGrouper(context, componentId, streamId, targetTasks, topoConf);
+            case ALL:
+                return new Grouper() {
+                    @Override
+                    public List<Integer> outTasks(Integer taskId, List<Object> values, LoadMapping load) {
+                        return targetTasks;
+                    }
+                };
+            case LOCAL_OR_SHUFFLE:
+                Set<Integer> sameTasks = Sets.intersection(Sets.newHashSet(targetTasks), Sets.newHashSet(context.getThisWorkerTasks()));
+                if (sameTasks.isEmpty()) {
+                    return new Grouper.ShuffleGrouper(context, componentId, streamId, targetTasks, topoConf);
+                } else {
+                    return new Grouper.ShuffleGrouper(context, componentId, streamId, new ArrayList<>(sameTasks), topoConf);
+                }
+            case NONE:
+                return new Grouper() {
+                    @Override
+                    public List<Integer> outTasks(Integer taskId, List<Object> values, LoadMapping load) {
+                        int index = random.nextInt(numTasks);
+                        return Collections.singletonList(targetTasks.get(index));
+                    }
+                };
+            case CUSTOM_OBJECT:
+                CustomStreamGrouping customObject =
+                    (CustomStreamGrouping) Thrift.instantiateJavaObject(thriftGrouping.get_custom_object());
+                return new Grouper.CustomGrouper(customObject, context, componentId, streamId, targetTasks);
+            case CUSTOM_SERIALIZED:
+                CustomStreamGrouping
+                    customSerialized =
+                    Utils.javaDeserialize(thriftGrouping.get_custom_serialized(), CustomStreamGrouping.class);
+                return new Grouper.CustomGrouper(customSerialized, context, componentId, streamId, targetTasks);
```

Comment:
If someone created a CustormStreamGrouping that was also LoadAwareCustomStreamGrouping the LoadAware portion would never be called, because we are now using constructors instead of the builder pattern.  This really should be.

```
case CUSTOM_SERIALIZED:
  CustomStreamGrouping
    customGrouping =
      Utils.javaDeserialize(thriftGrouping.get_custom_serialized(), CustomStreamGrouping.class);
    if (customGrouping instanceof CustomLoadAwareStreamGrouping) {
      return new Grouper.LoadAwareCustomGrouper(customSerialized, context, componentId, streamId, targetTasks);
    }
    return new Grouper.CustomGrouper(customSerialized, context, componentId, streamId, targetTasks);
```